FireboltPrivacy/FireboltPrivacyImplementation.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2020 Metrological
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Module.h"
+#include <interfaces/IConfiguration.h>
+#include <interfaces/IFireboltPrivacy.h>
+
+namespace WPEFramework {
+
+namespace Plugin
+{
+    class FireboltPrivacyImplementation : public Exchange::IFireboltPrivacy,
+                                         public Exchange::IConfiguration {
+
+    private:
+        class Config : public Core::JSON::Container {
+        public:
+            Config(const Config&) = delete;
+            Config& operator=(const Config&) = delete;
+
+            Config()
+                : Core::JSON::Container()
+                , devicemanifest("")
+                , fireboltconfigpath("")
+            {
+                Add(_T("devicemanifest"), &devicemanifest);
+                Add(_T("fireboltconfigpath"), &fireboltconfigpath);
+            }
+            ~Config() override = default;
+        public:
+            Core::JSON::String devicemanifest;
+            // This path is specific to firebolt. It points to the path where
+            // firebolt config files are placed. These files are shared between Ripple and other FireboltThunderPlugins
+            // In prod it points to /etc/
+            // and in dev it points to /opt/
+            Core::JSON::String fireboltconfigpath;
+        };
+
+
+    public:
+        FireboltPrivacyImplementation(const FireboltPrivacyImplementation&) = delete;
+        FireboltPrivacyImplementation& operator= (const FireboltPrivacyImplementation&) = delete;
+        FireboltPrivacyImplementation()
+            : _adminLock()
+            , _config()
+            , _deviceManifest()
+            , _allowResumePoints(false)
+        {
+        }
+        ~FireboltPrivacyImplementation() override
+        {
+            _deviceManifest.clear();
+            if (_service) {
+                _service->Release();
+                _service = nullptr;
+            }
+        }
+
+        uint32_t Register(Exchange::IFireboltPrivacy::INotification* notification) override
+        {
+            ASSERT(notification);
+            _adminLock.Lock();
+            notification->AddRef();
+            _notifications.push_back(notification);
```

Comment:
we also always have an assert that checks if you do not register the same notification twice (as that can lead to nasty bugs)